### PR TITLE
ImmutableCache: add redis support

### PR DESCRIPTION
**Pull request by mattnibs**
Support using redis as an lru cache for immutable files. Like local immutable
cache this current only supports lookups for chunk metadata files- this will be
extended in future iterations.

The next pr will include helm support for a service.

Part of #1899

### CODE REVIEW

File: ppl/zqd/redis.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	Enabled bool
+}
+
+func (c *RedisConfig) SetFlags(fs *flag.FlagSet) {
+	fs.Var(c, "redis.url", "redis connection url (i.e. redis://<user>:<password>@<host>:<port>/<db_number>)")
+	fs.BoolVar(&c.Enabled, "redis.enabled", false, "enable use of a redis server")
+	fs.StringVar(&c.Addr, "redis.addr", "localhost:6379", "redis address")
+	fs.StringVar(&c.Username, "redis.user", "", "redis username")
+	fs.StringVar(&c.Password, "redis.password", "", "redis password")
```
Comment (alfred-landrum) [inline ppl/zqd/redis.go:33]: Would you please add a `PasswordFile` option, like you did with postgres?

File: ppl/archive/immcache/immcache.go [hunk 2]
```diff
@@ -5,29 +5,98 @@ package immcache
...
+	// value of RedisKeyExpiration should not be set to 0.
+	RedisKeyExpiration cli.Duration
 }
 
 func (c *Config) SetFlags(fs *flag.FlagSet) {
-	fs.IntVar(&c.LocalCacheSize, "immcache.localsize", 0, "enables local caching of up to N small files (disabled if 0)")
+	c.RedisKeyExpiration = cli.Duration(time.Hour * 24)
+	fs.Var(&c.Kind, "immcache.kind", "kind of immutable cache")
+	fs.IntVar(&c.LocalCacheSize, "immcache.local.size", 128, "number of small files to keep in local cache")
+	fs.Var(&c.RedisKeyExpiration, "immcache.redis.keyexpiry", "expiration duration of immutable keys")
```
Comment (alfred-landrum) [inline ppl/archive/immcache/immcache.go:66]: There is a `fs.DurationVar`, is there a reason it wouldn't work here?

File: ppl/archive/immcache/immcache.go [hunk 3]
```diff
@@ -5,29 +5,98 @@ package immcache
...
+	Kind Kind
 	// LocalCacheSize specifies the number of immutable files to keep in a
 	// local lru cache used to speed up searches. Values less than or equal to 0
 	// (default), disables local caching of immutable files.
 	LocalCacheSize int
+	// RedisKeyExpiration is the time redis will keep the key of the immutable
+	// file around before deleting this. If RedisKeyExpiration == 0 the key will
+	// be kept around indefinitely. Because this is supposed to be used as an
+	// LRU cache, if the redis server eviction policy is set volitale-lru the
+	// value of RedisKeyExpiration should not be set to 0.
```
Comment (alfred-landrum) [inline ppl/archive/immcache/immcache.go:58]:
```suggestion
	// RedisKeyExpiration is the expiration value used when creating keys.
	// A value of zero (meaning no expiration) should only be used when
	// Redis is configured with a key eviction policy.
```

File: ppl/archive/immcache/redis.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/ppl/archive/chunk"
+	"github.com/go-redis/redis/v8"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type RedisCache struct {
+	metrics
+	expiry time.Duration
+	client *redis.Client
```
Comment (alfred-landrum) [inline ppl/archive/immcache/redis.go:18]:
```suggestion
	client *redis.Client
	expiry time.Duration
```

File: ppl/archive/immcache/immcache.go (hunk 2 above)
Comment (mattnibs) [reply ppl/archive/immcache/immcache.go:66]: I did not know that!

File: ppl/zqd/redis.go (hunk 1 above)
Comment (mattnibs) [reply ppl/zqd/redis.go:33]: I would prefer adding this in the next pr when that option will actually be used

File: ppl/archive/immcache/redis.go (hunk 4 above)
Comment (mattnibs) [reply ppl/archive/immcache/redis.go:18]: Whoops

**Review by alfred-landrum — APPROVED** [review]

**Pull request merged**